book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+|----------------------|-----------------------|-------------------------------------|
+| `hash_by_height`     | `BE32(height)`        | `block::Hash`                       |
+| `height_by_hash`     | `block::Hash`         | `BE32(height)`                      |
+| `block_by_height`    | `BE32(height)`        | `Block`                             |
+| `tx_by_hash`         | `transaction::Hash`   | `BE32(height) || BE32(tx_index)`    |
+| `utxo_by_outpoint`   | `OutPoint`            | `TransparentOutput`                 |
+| `sprout_nullifiers`  | `sprout::Nullifier`   | `()`                                |
+| `sapling_nullifiers` | `sapling::Nullifier`  | `()`                                |
+| `sprout_anchors`     | `sprout::tree::Root`  | `()`                                |
+| `sapling_anchors`    | `sapling::tree::Root` | `()`                                |
+
+Zcash structures are encoded using `ZcashSerialize`/`ZcashDeserialize`.
+
+### Notes on Sled trees
+
+- The `hash_by_height` and `height_by_hash` trees provide the bijection between
+  block heights and block hashes.  (Since the Sled state only stores finalized
+  state, this is actually a bijection).
+
+- Blocks are stored by height, not by hash.  This has the downside that looking
+  up a block by hash requires an extra level of indirection.  The upside is
+  that blocks with adjacent heights are adjacent in the database, and many
+  common access patterns, such as helping a client sync the chain or doing
+  analysis, access blocks in (potentially sparse) height order.  In addition,
+  the fact that we commit blocks in order means we're writing only to the end
+  of the Sled tree, which may help save space.
+
+- Transaction references are stored as a `(height, index)` pair referencing the
+  height of the transaction's parent block and the transaction's index in that
+  block.  This would more traditionally be a `(hash, index)` pair, but because
+  we store blocks by height, storing the height saves one level of indirection.
+
+## Committing finalized blocks
+
+If the parent block is not committed, add the block to an internal queue for
+future processing.  Otherwise, commit the block described below, then
+commit any queued children.  (Although the checkpointer generates verified
+blocks in order when it completes a checkpoint, the blocks are committed in the
+response futures, so they may arrive out of order).
+
+Committing a block to the sled state should be implemented as a wrapper around
+a function also called by [`Request::CommitBlock`](#request-commit-block),
+which should:
+
+1. Obtain the highest entry of `hash_by_height` as `(old_height, old_tip)`.
+Check that `block`'s parent hash is `old_tip` and its height is
+`old_height+1`, or panic. This check is performed as defense-in-depth to
+prevent database corruption, but it is the caller's responsibility (e.g. the
+zebra-state service's responsibility) to commit finalized blocks in order.
+
```

Comment:
The genesis block does not have a parent block. For genesis blocks,
check that `block`'s parent hash is `null` (all zeroes) and its height is `0`.